### PR TITLE
Added 'buttonset' component with modal support

### LINKED ISSUE #781: Button: Complete Api By Adding EP changes to NG
**Is your feature request related to a problem? Please describe.**
We added a new button api https://github.com/infor-design/enterprise-ng/issues/467 but its not fully coded into the NG project.

**Describe the solution you'd like**
- add the types
- show an example of modifying a modal dynamically in the page
- tests?

**Describe alternatives you've considered**
N/A

**Additional context**
Added 4.27

### PR DESCRIPTION
**Pull request by bthharper**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**

Provide access to the `buttonset` api added to the soho enterprise widgets.  This allows developers to interact with the buttons present on a modal dialog (or a `buttonset` added elsewhere).  

See https://github.com/infor-design/enterprise/pull/3582 for a full rationale for this change.

The API is exposed by the modal ref, and can be used in the dialog events, or once the dialog has been opened elsewhere.

Took the opportunity to add a few extra typings for `soho-button`, which now has `disabled`, and `pressed` properties.  In addition to `toggle` and `updated` methods.  The `element` property is also exposed.

**Related github/jira issue (required)**:

See #467. 
Closes #781 
 
**Steps necessary to review your pull request (required)**:

Git clone the **ids-enterprise-ng** repository.
```sh
npm i
npm run build
npm run lint
npm run test:lib
npm run start
```
Navigate to http://localhost:4200/ids-enterprise-ng-demo/buttonset
Exercise the generic button api.

http://localhost:4200/ids-enterprise-ng-demo/modal-dialog
Open the *Full On Tablet* modal dialog.
The button api used to disable/enable buttons.

### CODE REVIEW

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts [hunk 1]
```diff
@@ -5,26 +5,97 @@
...
   settings: SohoButtonOptions;
+
+  /** Associated HTML element. */
+  element: HTMLButtonElement;
+
+  /** Destroys any resources held by the button. */
   destroy(): void;
+
+  /** Returns true if the button is disabled; otherwise false. */
+  get disabled(); boolean;
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:71]: semi-colon after ()

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts [hunk 2]
```diff
@@ -5,26 +5,97 @@
...
 
-  
   /** Hides menu arrow usually displayed to the right of a menu button/icon. */
   hideMenuArrow?: boolean;
+
+  /** button style: defaults to "default". */
+  style?: SohoButtonOptionsStyle;
+
+  /** button type: defaults to "default". */
+  type?: SohoButtonOptionsType,
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:35]: line should end with semi-colon

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts [hunk 3]
```diff
@@ -5,26 +5,97 @@
...
+
+  /** Ripple */
+  ripple?: boolean,
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:38]: line should end with semi-colon

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts [hunk 4]
```diff
@@ -5,26 +5,97 @@
...
+
+  /** validate? */
+  validate?: boolean
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:41]: line should end with semi-colon

File: projects/ids-enterprise-ng/src/lib/buttonset/soho-buttonset.component.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+  @Input()
+  set detectHTMLButtons(value: boolean) {
+    this.settings.detectHTMLButtons = value;
+    if (this.buttonset) {
+      this.markForRefresh();
+    }
+  }
+
+  get detechHTMLButtons(): boolean {
+    if (this.buttonset) {
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/buttonset/soho-buttonset.component.ts:89]: should be detectHTMLButtons() - typo

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts [hunk 6]
```diff
@@ -5,26 +5,97 @@
...
+  element: HTMLButtonElement;
+
+  /** Destroys any resources held by the button. */
   destroy(): void;
+
+  /** Returns true if the button is disabled; otherwise false. */
+  get disabled(); boolean;
+
+  /** Sets the disable state of the button. */
+  set disabled(bal: boolean);
```
Comment (tmcconechy) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:74]: Whats bal? maybe val or value?

File: karma.conf.js [hunk 7]
```diff
@@ -3,34 +3,34 @@
...
     client: {
-      clearContext: false // leave Jasmine Spec Runner output visible in browser
+      clearContext: false, // leave Jasmine Spec Runner output visible in browser
     },
     coverageIstanbulReporter: {
-      dir: require('path').join(__dirname, '../coverage'),
-      reports: ['html', 'lcovonly', 'text-summary'],
-      fixWebpackSourcePaths: true
+      dir: require("path").join(__dirname, "../coverage"),
+      reports: ["html", "lcovonly", "text-summary"],
```
Comment (fitzorama) [inline karma.conf.js:20]: is this correct, changing single quotes to double?

File: src/app/buttonset/buttonset.demo.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<div class="row">
```
Comment (tmcconechy) [inline src/app/buttonset/buttonset.demo.html:1]:
```suggestion
<div class="row top-padding">
```

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts (hunk 6 above)
Comment (tmcconechy) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:74]:
```suggestion
  set disabled(value: boolean);
```

**Comment by EdwardCoyle** [comment]
@tmcconechy I'm not 100% the integration with ButtonSet/CAP works.  CAP uses a standard toolbar underneath instead of ButtonSet, so at the moment it won't be affected by this.  We should open a separate EP issue to review that.

@bthharper thank you for your work on this, aside from Tim/Mike's comments I think it looks great.

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts (hunk 2 above)
Comment (EdwardCoyle) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:35]:
```suggestion
  type?: SohoButtonOptionsType;
```

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts (hunk 1 above)
Comment (EdwardCoyle) [inline projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:71]:
```suggestion
  get disabled(): boolean;
```

File: projects/ids-enterprise-ng/src/lib/buttonset/soho-buttonset.component.ts [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+   */
+  @Input()
+  set detectHTMLButtons(value: boolean) {
+    this.settings.detectHTMLButtons = value;
+    if (this.buttonset) {
+      this.markForRefresh();
+    }
+  }
+
+  get detechHTMLButtons(): boolean {
```
Comment (EdwardCoyle) [inline projects/ids-enterprise-ng/src/lib/buttonset/soho-buttonset.component.ts:88]:
```suggestion
  get detectHTMLButtons(): boolean {
```

**Comment by tmcconechy** [comment]
OK thanks @EdwardCoyle wasnt sure if it was just going to work or not made https://github.com/infor-design/enterprise/issues/3951

I will approve then when the fixes above are done.

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts (hunk 1 above)
Comment (bthharper) [reply projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts:71]: fixed - not sure why lint doesn't pick these up.

File: karma.conf.js (hunk 7 above)
Comment (bthharper) [reply karma.conf.js:20]: I think VS Code did a format on it,  I'll try to revert.

**Comment by tmcconechy** [comment]
@bthharper still a couple in here i wasnt sure about in the comments..

File: projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts (hunk 2 above)
Comment (bthharper) [reply projects/ids-enterprise-ng/src/lib/button/soho-button.d.ts]: Fixed.

**Pull request merged**